### PR TITLE
Add compilation of complex pattern matching to case

### PR DESCRIPTION
**Pull request by paulcadman**
This PR adds the `match-to-case` Core transformation. This transforms pattern matching nodes to a sequence of case and let nodes.

## High level description

Each branch of the match is compiled to a lambda. In the combined match 

Each branch of the match is compiled to a lambda. These lambdas are combined in nested lets and each lambda is called in turn as each branch gets checked. The lambda corresponding to the first branch gets called first, if the pattern match in the branch fails, the lambda corresponding to the next branch is called and so on. If no branches match then a lambda is called which returns a fail node.

Conceptually:

<table>
<tr>
<td>
Core
</td>
<td>
Transformed
</td>
</tr>
<tr>
<td>

```

match v1 .. vn {
  b1
  b2
  ...
  bk
}

```

</td>
<td>

```
λ
  let c0 := λ FAIL in
    let ck := λ {...} in
      ...
      let c1 := λ {...} in
  c1 v1 ... vn

```

</td>
</tr>
</table>

The patterns on each branch are compiled to either let bindings (pattern
binders) or case expressions (constructor patterns).

Auxillary bindings are added in the case of nested constructor patterns.

The default branch in each case expression has a call to the lambda
corresponding to the next branch of the match. This is because the default
branch is reached if the pattern match fails.

<table>

<tr>
<td>
Pattern match
</td>
<td>
Transformed
</td>
</tr>
<tr>
<td>

```
suc (suc n) ↦ n
```

</td>
<td>

```
  case ?$0 of {
    suc arg_8 := case ?$0 of {
      suc n := let n := ?$0 in n$0;
      _ := ?$2 ?$1
    };
    _ := ?$1 ?$0
  }

```

</td>
</tr>
</table>

The body of each branch is wrapped in let bindings so that the indicies of bound
variables in the body point to the correct variables in the compiled expression.
This is necessary because the auxiliary bindings added for nested constructor
patterns will cause the original indicies to be offset.

Finally, the free variables in the match branch body need to be shifted by all the bindings we've added as part of the compilation.

### Single wildcard

<table>
<tr>
<td> Juvix </td> <td> Core </td> <td> Transformed Core </td>
</tr>
<tr>
<td>

```
f : Nat -> Nat;
f _ := 1;
```

</td>
<td>

```
λ? match ?$0 with {
  _ω309 ↦ ? 1
}
```

</td>
<td>

```
λ? let ? := λ? fail "Non-exhaustive patterns" in
   let ? := λ? let _ω309 := ?$0 in
               let _ω309 := ?$0 in 1 in
   ?$0 ?$2
```

</td>
</tr>
</table>

### Single binder

<table>
<tr>
<td> Juvix </td> <td> Core </td> <td> Transformed Core </td>
</tr>
<tr>
<td>

```
f : Nat -> Nat;
f n := n;
```

</td>
<td>

```
λ? match ?$0 with {
  n ↦ n$0
}
```

</td>
<td>

```
λ? let ? := λ? fail "Non-exhaustive patterns" in
   let ? := λ? let n := ?$0 in
               let n := ?$0 in n$0 in
   ?$0 ?$2
```

</td>
</tr>
</table>

### Single Constructor

<table>
<tr>
<td> Juvix </td> <td> Core </td> <td> Transformed Core </td>
</tr>
<tr>
<td>

```
f : Nat -> Nat;
f (suc n) := n;
```

</td>
<td>

```
λ? match ?$0 with {
  suc n ↦ n$0
}
```

</td>
<td>

```
λ? let ? := λ? fail "Non-exhaustive patterns" in let ? := λ? case ?$0 of {
  suc n := let n := ?$0 in let n := ?$0 in n$0;
  _ := ?$1 ?$0
} in ?$0 ?$2
```

</td>
</tr>
</table>

### Nested Constructor

<table>
<tr>
<td> Juvix </td> <td> Core </td> <td> Transformed Core </td>
</tr>
<tr>
<td>

```
f : Nat -> Nat;
f (suc (suc n)) := n;
```

</td>
<td>

```
λ? match ?$0 with {
  suc (suc n) ↦ n$0
}
```

</td>
<td>

```
λ? let ? := λ? fail "Non-exhaustive patterns" in let ? := λ? case ?$0 of {
  suc arg_8 := case ?$0 of {
    suc n := let n := ?$0 in let n := ?$0 in n$0;
    _ := ?$2 ?$1
  };
  _ := ?$1 ?$0
} in ?$0 ?$2
```

</td>
</tr>
</table>

### Multiple Branches

<table>
<tr>
<td> Juvix </td> <td> Core </td> <td> Transformed Core </td>
</tr>
<tr>
<td>

```
f : Nat -> Nat;
f (suc n) := n;
f zero := 0;
```

</td>
<td>

```
λ? match ?$0 with {
  suc n ↦ n$0;
  zero ↦ ? 0
}
```

</td>
<td>

```
λ? let ? := λ? fail "Non-exhaustive patterns" in let ? := λ? case ?$0 of {
  zero := ? 0;
  _ := ?$1 ?$0
} in let ? := λ? case ?$0 of {
  suc n := let n := ?$0 in let n := ?$0 in n$0;
  _ := ?$1 ?$0
} in ?$0 ?$3
```

</td>
</tr>
</table>

### Nested case with captured variable

<table>
<tr>
<td> Juvix </td> <td> Core </td> <td> Transformed Core </td>
</tr>
<tr>
<td>

```
f : Nat -> Nat -> Nat;
f n m := case m
  | suc k := n + k;
```

</td>
<td>

```
f = λ? λ? match ?$1, ?$0 with {
  n, m ↦ match m$0 with {
    suc k ↦ + n$2 k$0
  }
}
```

</td>
<td>

```
λ? λ?
  let ? := λ? λ? fail "Non-exhaustive patterns" in
  let ? := λ? λ? let n := ?$1 in let m := ?$1 in let n := ?$1 in let m := ?$1 in
      let ? := λ? fail "Non-exhaustive patterns" in let ? := λ? case ?$0 of {
            suc k := let k := ?$0 in let k := ?$0 in + n$6 k$0;
            _ := ?$1 ?$0
} in ?$0 m$2 in ?$0 ?$3 ?$2
```

</td>
</tr>
</table>

## Testing

The `tests/Compilation/positive` tests are run up to the Core evaluator with `match-to-case` and `nat-to-int` transformations on Core turned on.

### CODE REVIEW

**Comment by lukaszcz** [comment]
Why does the merge sort test fail?

**Comment by paulcadman** [comment]
> Why does the merge sort test fail?

I didn't notice that it was filtered, it was left filtered out by https://github.com/anoma/juvix/pull/1740 so I haven't run it. I can check.

**Comment by lukaszcz** [comment]
> > Why does the merge sort test fail?
> 
> I didn't notice that it was filtered, it was left filtered out by #1740 so I haven't run it. I can check.

It seems to reveal a bug in the match-to-case translation, because it works fine in the REPL without the translation.

**Comment by paulcadman** [comment]
> > > Why does the merge sort test fail?
> > 
> > 
> > I didn't notice that it was filtered, it was left filtered out by #1740 so I haven't run it. I can check.
> 
> It seems to reveal a bug in the match-to-case translation, because it works fine in the REPL without the translation.

You're right - the translation fails, specifically for the `uniq` function in that file. I will investigate/fix. Thanks.

**Comment by lukaszcz** [comment]
It may be worthwhile to test the match-to-test transformation on the Core level. For this purpose, I extended the JuvixCore test for the trans. Now the test for match expressions is failing.

**Comment by lukaszcz** [comment]
It may be worthwhile to test the match-to-test transformation on the Core level. For this purpose, I extended the JuvixCore test for the transformation pipeline with the MatchToCase transformation at the beginning (this test runs the transformations and checks if the evaluation results are preserved). Now the test for match expressions is failing.

**Comment by lukaszcz** [comment]
I isolated an error to a single case expression. There seems to be some problem with variables inside nested patterns. I added a test.

**Review by lukaszcz — APPROVED** [review]
Great we have something working! Doing it line-by-line and creating a new function for every pattern is very inefficient, but this will be fixed in the future by implementing the Wadler algorithm.

File: tests/Compilation/positive/out/test035.out [hunk 1]
```diff
@@ -0,0 +1,7 @@
+9 :: 7 :: 5 :: 3 :: 1 :: nil
+280
+571200
+7896000
+45289200
+375572736
+5112834048000
```
[the remark this answers is not in the document]
Comment (paulcadman) [reply tests/Compilation/positive/out/test035.out]: I modified `Core/positive/test041.jvc` to get these results

**Comment by jonaprieto** [comment]
Great!

**Pull request merged**